### PR TITLE
Enable post upload script to work with non-published datasets

### PR DESCRIPTION
**Pull request by pli888**
# Pull request for issue: #1664

This is a pull request for the following functionalities:

* The `md5.sh` script can now create a `*.filesizes` file containing sizes of files for a dataset, in addition to the `doi.md5` file. Both md5 and filesizes files are then uploaded into a S3 bucket called `gigadb-datasets-metadata`.
* The `FilesCommand` class fetches the `doi.md5` file from S3 `gigadb-datasets-metadata` bucket for updating md5 attributes.
* File Metadata Console tool to update dataset file sizes in database using information stored in `doi.filesizes` file that's located in S3 bucket.
* Post upload script can update md5 values and file sizes for non-published and published datasets using above updated tools. The postUpload.sh script has been fixed to upload readme files into Wasabi.

## How to test?

Create a feature branch in your repo containing the code in the pull request:
```
$ git checkout -b pli888-postupload-script-improvements develop
$ git pull https://github.com/pli888/gigadb-website.git postupload-script-improvements
```

#### Dev environment

Testing the latest functionality in `md5.sh` in this pull request requires the [AWS command line tool](https://aws.amazon.com/cli/) installed on your environment. If you're using homebrew, this can be done using `brew install awscli`. You will then need to configure it to access AWS resources using your personal IAM user credentials by creating `~/.aws/config` and `~/.aws/credentials` files.

To manually test `md5.sh` on your dev environment mirrors, we mirror how curators use the current script.
```
# Change directory to a dataset that you want to create a md5 file for
$ cd gigadb/app/tools/files-metadata-console/tests/_data/102480
# Execute md5.sh
$ ../../../scripts/md5.sh
Created 102480.md5
Created 102480.filesizes
upload: ./102480.filesizes to s3://gigadb-datasets-metadata/102480.filesizes
upload: ./102480.md5 to s3://gigadb-datasets-metadata/102480.md5  
```

You will see 2 new files have been created in the dataset's directory: `102480.filesizes` and `102480.md5`.
```
# 102480.md file contains md5 checksums of 2 files
$ more 102480.md5 
1b31864478eec0479ba76ff242cd7dbc  ./readme_102480.txt
67d9336ca3b61384185dc665026a2325  ./analysis_data/Tree_file.txt

# 102480.filesizes contains size of 2 files in bytes
$ more 102480.filesizes 
    3202        ./readme_102480.txt
     359        ./analysis_data/Tree_file.txt
```

Your AWS IAM user accounts have access to the S3 bucket `gigadb-datasets-metadata`. A new S3 policy has been created called `AllowReadWriteBucketGigadbDatasetsMetadata` based on `policy-s3.md` and attached to IAM `Developers` group. If you check the [contents](https://s3.console.aws.amazon.com/s3/buckets/gigadb-datasets-metadata?region=ap-northeast-1&bucketType=general&tab=objects) of this bucket, you will see that the md5 and filesizes files have been uploaded into it.

#### Live GigaDB file server environment

The new `md5.sh` located in `/home/peter/md5.sh` has been checked to work on the live GigaDB file server which already has AWS CLI installed. This tool will use AWS credentials for a new temporary IAM user called `ExcelFileUpload` who only has permission to upload files to the S3 `gigadb-datasets-metadata` bucket using the `AllowReadWriteBucketGigadbDatasetsMetadata` policy. Its AWS access configuration can be found in `/etc/aws` directory.

#### Dev environment

There is a 100006.filesizes containing made-up data in S3 bucket. This can be used by the files-metadata-console tool to update the missing file sizes for dataset http://gigadb.gigasciencejournal.com/dataset/100006.
```
# Spin up local GigaDB
$ ./up.sh
# Confirm missing file sizes in http://gigadb.gigasciencejournal.com/dataset/100006
$ cd gigadb/app/tools/files-metadata-console
$ docker-compose run --rm files-metadata-console ./yii update/file-sizes --doi=100006
# Restart web server container service if you cannot see new file sizes in 100006
$ cd ../../../..
$ docker compose restart web
```

#### Staging environment

Spin up GigaDB on your staging server:
```
$ cd /path/to/gigadb-website/ops/infrastructure/envs/staging

# Copy terraform files to staging environment
$ ../../../scripts/tf_init.sh --project gigascience/forks/pli888-gigadb-website --env staging

# Provision with Terraform 
$ terraform plan
$ terraform apply
$ terraform refresh

# Copy ansible files
$ ../../../scripts/ansible_init.sh --env staging

# Provision bastion ec2 server and restore RDS with latest database backup
$ env OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook -i ../../inventories bastion_playbook.yml --extra-vars="gigadb_env=staging"

# Provision webapp ec2 server with ansible
$ TF_KEY_NAME=private_ip env OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook -i ../../inventories webapp_playbook.yml --extra-vars="gigadb_env=staging"
```

Run CI/CD pipeline with code from this feature branch all the way to `sd_deploy` job.

Take a look at the https://your-staging-gigadb/dataset/100006 page and make a note of its md5 values, especially for the readme file.

SSH into your staging bastion server and update the file size values for dataset 100006:
```
$ terraform output
$ ssh -i ~/.ssh/your-pem-file.pem centos@staging-bastion-ip
[centos@ip-10-99-0-188 ~]$ docker run --rm "registry.gitlab.com/$GITLAB_PROJECT/production-files-metadata-console:$GIGADB_ENV" ./yii update/file-sizes --doi=100006
Number of changes: 7
```

Take another look at the https://your-staging-gigadb/dataset/100006 page and see if it has changed.

#### Dev environment

Test `FilesCommand` to update md5 values from `*.md5` files stored in the `gigadb-datasets-metadata` S3 bucket.
```
$ docker-compose run --rm  test ./protected/yiic files updateMD5FileAttributes --doi=100006
```

If you now go to http://gigadb.gigasciencejournal.com/dataset/100006 in your browser, you should see the MD5 values of the dataset have been updated from a 100006.md file from the S3 bucket. You can confirm this by looking at the md5 value for the readme.txt file which should have a made-up value.

#### Staging environment

Take a look at the https://your-staging-gigadb/dataset/100006 page and make a note of its md5 values, especially for the readme file.

SSH into your staging bastion server and update the md5 values for dataset 100006:
```
$ ssh -i ~/.ssh/your-pem-file.pem centos@staging-bastion-ip
[centos@ip-10-99-0-188 ~]$ docker run -e YII_PATH=/var/www/vendor/yiisoft/yii "registry.gitlab.com/$GITLAB_PROJECT/production_app:$GIGADB_ENV" ./protected/yiic files updateMD5FileAttributes --doi=100006
Saved md5 file attribute with id: 10674
Saved md5 file attribute with id: 10672
Saved md5 file attribute with id: 10673
Saved md5 file attribute with id: 10671
Saved md5 file attribute with id: 10670
Saved md5 file attribute with id: 10669
Saved md5 file attribute with id: 10675
```

Now, check your https://your-staging-gigadb/dataset/100006 page again and you will see that the md5 values have been updated which will be confirmed by the md5 value of the readme.txt file.

#### Dev environment

Reset the database for your GigaDB dev environment:
```
$ pwd
/path/to/gigadb-website
$ ./ops/scripts/setup_devdb.sh dev
```

Go to http://gigadb.gigasciencejournal.com/dataset/100006 on your browser and make a note of the contents of the Files tab for dataset 100006. You will see that there are no md5 checksum values and file sizes are as expected.

Now, execute the post upload script:
```
$ cd gigadb/app/tools/excel-spreadsheet-uploader
$ ./postUpload.sh 100006
```

Refresh the dataset page for 100006. If you look in the Files tab, you should see the dataset files have md5 checksum values and the sizes of the files have been updated. This tells you that the `postUpload.sh` script has successfully executed. Further confirmation can be seen by looking at the output files in readme-generator/runtime/curators directory.

#### Staging environment

On your staging server, make a change to the size of a file in dataset 100006 using the admin file page. By now running postUpload.sh, this script will revert the change you have just made confirming it works for published and non-published datasets because it uses md5 and filesizes files stored in a generic location provided by a S3 bucket

> [!NOTE]  
> If you run the postUpload.sh script again, you should not see any delays with updating file sizes because database triggers are dropped before changes to the database are made. This suggests the delays with updating file sizes reported in #1662 maybe due to the presence of database triggers.

SSH into your bastion staging server and execute post upload script as follows:
```
[centos@ip-10-99-0-188 ~]$ ./postUpload.sh 100006

* About to update files' size for 100006
Number of changes: 7

Done with updating files' size for 100006. Nb of successful changes saved in file: /home/centos/uploadLogs/updating-file-size-100006.txt

* About to update files' MD5 Checksum as file attribute for 100006
Saved md5 file attribute with id: 10674
Saved md5 file attribute with id: 10672
Saved md5 file attribute with id: 10673
Saved md5 file attribute with id: 10671
Saved md5 file attribute with id: 10670
Saved md5 file attribute with id: 10669
Saved md5 file attribute with id: 10675

Done with updating files' MD5 Checksum as file attribute for 100006. Process status is saved in file: /home/centos/uploadLogs/updating-md5checksum-100006.txt
[DOI] 10.5524/100006

[Title] Genomic data from Adelie penguin (<em>Pygoscelis adeliae</em>). 

[Release Date] 2011-07-06

[Citation] Wang, J; Zhang, G; Lambert, DM (2011): Genomic data from Adelie
penguin (<em>Pygoscelis adeliae</em>).  GigaScience Database.
https://dx.doi.org/10.5524/100006

[Data Type] Genomic

[Data Summary] The Adelie penguin (<em>Pygoscelis adeliae</em>) is an iconic
penguin of moderate stature and a tuxedo of black and white feathers.  The
penguins are only found in the Antarctic region and surrounding islands.  Being
very sensitive to climate change, and due to changes in their behavior based on
minor shifts in climate, they are often used as a barometer of the Antarctic.
<br>
With its status as one of the adorable and cuddly flightless birds of
Antarctica, they serve as an example for conservation, and as a result they are
now categorised at low risk for endangerment.  The sequence of the penguin can
be of use in understanding the genetic underpinnings of its evolutionary traits
and adaptation to its extreme environment; its unique system of feathers; its
prowess as a diver; and its sensitivity to climate change.  We hope that this
genome data will further our understanding of one of the most remarkable
creatures to waddle the planet Earth.<br> We sequenced the genome of an adult
male from Inexpressible Island, Ross Sea, Antartica (provided by David Lambert)
to a depth of approximately 60X with short reads from a series of libraries with
various insert sizes (200bp- 20kb). 
The assembled scaffolds of high quality sequences total 1.23 Gb, with the contig
and scaffold N50 values of 19 kb and 5 mb respectively. We identified 15,270
protein-coding genes with a mean length of  21.3 kb.


[File Location] https://ftp.cngb.org/pub/gigadb/pub/10.5524/100001_101000/100006/

[File name] - [File Description]
Pygoscelis_adeliae.scaf.fa.gz  -  
Pygoscelis_adeliae.gff.gz  -  coding sequence annotation of assembly
Pygoscelis_adeliae.pep.gz  -  peptide translations of CDS predictions
Pygoscelis_adeliae.cds.gz  -  coding sequence predictions on genome assembly
Pygoscelis_adeliae.RepeatMasker.out.gz  -  repeat masker results
readme.txt  -  
Pygoscelis_adeliae.fa.gz  -  assembled scaffolds from sequence data

[License]
All files and data are distributed under the CC0 1.0 Universal (CC0 1.0) Public
Domain Dedication (https://creativecommons.org/publicdomain/zero/1.0/), unless
specifically stated otherwise, see http://gigadb.org/site/term for more details.

[Comments]

[End]
Created readme file and uploaded it to Wasabi gigadb-website/staging bucket directory

All postUpload logs have been moved to: /home/centos/uploadDir

PostUpload jobs done!
```

Try running the postUpload.sh script again with DOIs 102480 and 102499. No hanging delays when running the script should be seen.

## How have functionalities been implemented?

The new `md5.sh` script has functionality to create `doi.filesizes` file containing sizes of files for a dataset. It also uploads `doi.md5` and `doi.filesizes` files into a specified S3 bucket. Developers and users on the GigaDB file server can execute `md5.sh` to upload files into S3 bucket because of permissions from a S3 policy attached to them.

The underlying tools used by the post upload script have been updated to work with non-published datasets in a way that retains current functionality. In files metadata console, there is a new function in `DatasetFilesUpdater` component class called `updateFileSizes()` which will read the file size information for a dataset in a `doi.filesizes` file from a S3 bucket called `gigadb-datasets-metadata` and use it to update the file sizes for a dataset in the GigaDB PostgreSQL database. There is a corresponding `actionFileSizes(`) function in files metadata console's `UpdateController` class which calls `updateFileSizes()` in `DatasetFilesUpdater`.

To enable updating of md5 values for both published and non-published datasets, the `findDatasetMd5FileUrl($dataset)` function in `FileCommand` class has been updated to look for `doi.md5` file in S3 gigadb-datasets-metadata bucket first before trying to find the `doi.md5` file in a public dataset FTP directory.

The `postUpload.sh` script has been updated so that it calls `updateFileSizes()` in the files metadata console's `UpdateController`. PostUpload.sh also uses the createReadme.sh script to create readme files on staging and live bastion servers which allows readme files to be uploaded to Wasabi. File URL validation in postUpload.sh has been skipped for now, since this will only work on published datasets, but we need the post upload script to work on non-published data files.

## Any issues with implementation?

None.

## Any changes to automated tests?

The data for `gigadb_testdata` has been updated due to the fact that dataset 100005 contains a file whose location is `100142/Diagram-ALL-FIELDS-Check-annotation.jpg`. To correct this discrepancy, the identifier for dataset 100005 has been changed to 100142 in `gigadb.testdata/dataset.csv`. Functional tests for files metadata console have been updated according to this change in gigadb_testdata. Also, unit and functional test in ReadmeCest had to be updated because of this change.

A bats core test has been designed for `md5.sh` which tests doi.md5 and doi.filesizes files are created and they contain expected data.

There is a unit test for the new `updateFileSizes()` function in `DatasetFilesUpdater` component class.

## Any changes to documentation?

files-metadata-console/README.md has been updated on how to run all functional and unit tests at once.

There is now a `TROUBLESHOOTING_EXCEL_UPLOADS.md` which contains helpful information on how to resolve problems when uploading Excel spreadsheet files.

`PRODUCTION_TROUBLESHOOT.md` has been updated to use manual controls on Gitlab pipelines page to teardown and spin up Docker containers and other resources when recreating TLS certificates.

### CODE REVIEW

File: gigadb/app/tools/files-metadata-console/tests/functional/CheckValidURLsCest.php [hunk 1]
```diff
@@ -16,6 +16,9 @@ class CheckValidURLsCest
         "https://ftp.cngb.org/pub/gigadb/pub/10.5524/100001_101000/100006/phylogeny_study_update/",
     ];
 
+    /**
+     * @skip Skipping this test due to unreliable resolution of URLs
+     */
```
Comment (kencho51) [inline gigadb/app/tools/files-metadata-console/tests/functional/CheckValidURLsCest.php]: This test is fixed in PR https://github.com/gigascience/gigadb-website/pull/1745/files#diff-fe4bd02276e47a91f99dc3db159df2b55286b8e6c910101bcf23700e86a6f643

File: gigadb/app/tools/files-metadata-console/tests/functional/UpdateFileSizeCest.php [hunk 2]
```diff
@@ -51,8 +51,17 @@ public function tryUpdateFileSizeWhenContentLengthInBytes(\FunctionalTester $I):
...
     }
+
+    public function tryUpdateFileSizesFromTsvFile(\FunctionalTester $I): void
+    {
+        $webClient = new Client([ 'allow_redirects' => false ]);
+        $us = new URLsService();
+        $dfu = new DatasetFilesUpdater(["doi" => "100142", "us" => $us, "webClient" => $webClient]);
+        $success = $dfu->updateFileSizes();
+//        $I->assertEquals(1, $success, "Not all files were updated successfully");
+    }
```
Comment (kencho51) [inline gigadb/app/tools/files-metadata-console/tests/functional/UpdateFileSizeCest.php]: **question:** This test is still WIP? This test seems duplicated with the tests in `DatasetFilesUpdaterTest.php`

**Review by kencho51 — COMMENTED** [review]
Hi @pli888,

**praise:** All updated tests are passing in dev and gitlab pipeline
**praise:** The files for file size and md5 values are stored in the s3 bucket, which is more reliable and stable
**praise:** Confirmed that the md5 value of the files are fetched from the files stored in the s3 bucket 
**suggestion:** After executing `./postUpload.sh 100006`in staging, all logs will be stored in `/home/$user/uploadDir`, execept the log for `createReadme.sh` which will be found in `/home/$user/logs/`, saving it to `/home/$user/uploadDir` will make the logs be found more easily
**suggestion:** Should include the policy `AllowReadWriteBucketGigadbDatasetsMetadata` to `docs/awsdocs/`
**suggestion(non-blocking):** The new function `updateFileSizes()` with name very similar with the existing `updateFileSize()`, is it better rename it, eg. `updateFileSizesFormS3()`, for better readability?

File: gigadb/app/tools/files-metadata-console/tests/functional/CheckValidURLsCest.php (hunk 1 above)
Comment (pli888) [reply gigadb/app/tools/files-metadata-console/tests/functional/CheckValidURLsCest.php]: I removed skip tag now.

File: gigadb/app/tools/files-metadata-console/tests/functional/UpdateFileSizeCest.php (hunk 2 above)
Comment (pli888) [reply gigadb/app/tools/files-metadata-console/tests/functional/UpdateFileSizeCest.php]: I agree. This test has been removed now.

**Comment by pli888** [comment]
@kencho51

>**suggestion:** After executing `./postUpload.sh 100006`in staging, all logs will be stored in `/home/$user/uploadDir`, execept the log for `createReadme.sh` which will be found in `/home/$user/logs/`, saving it to `/home/$user/uploadDir` will make the logs be found more easily

Ok, this has been implemented now.

>**suggestion:** Should include the policy `AllowReadWriteBucketGigadbDatasetsMetadata` to `docs/awsdocs/`

I've added this policy to the bottom of `docs/awsdocs/policy-s3.md`.

>**suggestion(non-blocking):** The new function `updateFileSizes()` with name very similar with the existing `updateFileSize()`, is it better rename it, eg. `updateFileSizesFormS3()`, for better readability?

I did think about using a more informative name for `updateFileSizes()` when I created it. But then I also documented where this function gets file size information in the PHPDoc comment at the top of the `updateFileSizes()` function in the source code.

**Review by kencho51 — APPROVED** [review]
Hi @pli888,

As all my concerns have been addressed, so I am happy to approve.

**Review by rija — COMMENTED** [review]
Hi @pli888,

**praise**: functionality works locally and on staging
**issue**: on staging, the script is slow, probably due to the trigger issues we talked about
**suggestion**: Like in `execute.sh` add calls to drop and recreate triggers around the DB updates
**praise**: all automated tests are passing locally and on CI
**issue**: `md5.bats` requires us to be in `gigadb/app/tools/files-metadata-console/tests/bats` directory to run, yet there is a `updateUrls.bats` test in `gigadb/app/tools/files-metadata-console/tests/`,
Which means that we don't have a way to run all the BATS tests of the project in one go.
Furthermore, the PHP tests can be run from `gigadb/app/tools/files-metadata-console/` with just `./test`
**suggestion**: Have all BATS tests in `gigadb/app/tools/files-metadata-console/tests/bats` but make sure we can run them from `gigadb/app/tools/files-metadata-console/` like the PHP tests, with `bats tests/bats`
**issue**: `$GIGADB_DATASETS_METADATA_BUCKET_URL` should  not be a variable (`public string`), but a constant (the URL is almost never going to change and constants take less memory than variable, and immutable-unless-stricly-necessary is a good principle to follow), especially that it is already all capitalised which is the PHP convention for constants
**suggestion**: replace `public string` with `public const` and remove the `$` prefix
**praise**: Updating of MD5 checksum and file sizes follow the same coherent good approach

**Comment by pli888** [comment]
@rija 

>issue: on staging, the script is slow, probably due to the trigger issues we talked about
suggestion: Like in execute.sh add calls to drop and recreate triggers around the DB updates

This has been implemented in eab17a1. Running `postUpload.sh` multiple times on my staging server does not cause execution to hang anymore but completes within 1 minute.

>issue: md5.bats requires us to be in gigadb/app/tools/files-metadata-console/tests/bats directory to run, yet there is a updateUrls.bats test in gigadb/app/tools/files-metadata-console/tests/, which means that we don't have a way to run all the BATS tests of the project in one go. Furthermore, the PHP tests can be run from gigadb/app/tools/files-metadata-console/ with just ./test
suggestion: Have all BATS tests in gigadb/app/tools/files-metadata-console/tests/bats but make sure we can run them from gigadb/app/tools/files-metadata-console/ like the PHP tests, with bats tests/bats

The above has been implemented so that running `bats tests/bats` will execute bats tests in updateUrls.bats and md5.bats. Also, executing `./test` will now run unit, functional and bats tests in `files-metadata-console` on `dev` environment. On the CI/CD pipeline, it will run `files-metadata-console` unit tests as well functional tests.

>issue: $GIGADB_DATASETS_METADATA_BUCKET_URL should not be a variable (public string), but a constant (the URL is almost never going to change and constants take less memory than variable, and immutable-unless-stricly-necessary is a good principle to follow), especially that it is already all capitalised which is the PHP convention for constants
suggestion: replace public string with public const and remove the $ prefix

This has been implemented in 8636335.

**Review by rija — APPROVED** [review]
Based on the updates, I'm happy to approve.

**Pull request merged**